### PR TITLE
Fix post-flatterning textures

### PR DESCRIPTION
**Pull request by nathanruiz**
This change updates the textures to use the 1.19.2 release. I've tested this on a number of versions, including 1.8.9, 1.12.2, 1.13.2, and 1.16.5. This covers pre and post flattening.

The largest change is the rework of the `define_blocks` macro. Rather than including `data`, `offset`, and `offsets` as a method of resolving the block ID, I've changed it to just store information about a block with different states, and then have a separate mapping for the IDs at different versions. The benefit of this is we can generate both the `define_blocks` block, and the mappings from information in https://github.com/PrismarineJS/minecraft-data. I've included a tool called `generate-blocks` that will generate those new structures.

For all post-flattening versions, this mapping is a const array of each block state in order. This will add ~1.4MiB to the compiled size, but will allow for very fast lookups from IDs. For the pre-flattening versions, I instead made a match block since it needs to do a lookup on both block ID and blockstate.

The pre-flattening file (`blocks/src/versions/legacy.rs`) was the only mapping file that can't currently be automatically updated. I generated the initial version from `minecraft-data`, but then needed to make manual changes to get it to work correctly. I've done lots of testing on 1.8.9 and 1.12.2 to try to catch any issues. While doing this, I found a few bugs in the original code, particularly around `update_state` that are now fixed.

After looking into the `update_state` structure, it seems like it is only required for pre-flattening versions, since post-flattening versions just send multiple block updates in those cases. A few cases I tested for this were top half of doors, chests/double chests, stairs, pumpkin stems, and redstone. Pre 1.13, some of these states aren't sent by the server, but are intended to be calculated on the client side. After 1.12, these states are calculated on the server side. Since it's no longer required in 1.13+, I've disabled calls to `update_state` on those versions.

### CODE REVIEW

File: blocks/src/versions/legacy.rs [hunk 1]
```diff
@@ -0,0 +1,1692 @@
...
+        (11, 6) => Some(Block::Lava { level: 6 }),
+        (11, 7) => Some(Block::Lava { level: 7 }),
+        (11, 8) => Some(Block::Lava { level: 8 }),
+        (11, 9) => Some(Block::Lava { level: 9 }),
+        (11, 10) => Some(Block::Lava { level: 10 }),
+        (11, 11) => Some(Block::Lava { level: 11 }),
+        (11, 12) => Some(Block::Lava { level: 12 }),
+        (11, 13) => Some(Block::Lava { level: 13 }),
+        (11, 14) => Some(Block::Lava { level: 14 }),
+        (11, 15) => Some(Block::Lava { level: 15 }),
```
Comment (terrarier2111) [inline blocks/src/versions/legacy.rs:103]: couldn't you make this (11, 0..=15)? and then make the level `id & 0xF` LLVM should be able to optimize this out.

File: blocks/src/versions/legacy.rs [hunk 2]
```diff
@@ -0,0 +1,1692 @@
...
+        (9, 6) => Some(Block::Water { level: 6 }),
+        (9, 7) => Some(Block::Water { level: 7 }),
+        (9, 8) => Some(Block::Water { level: 8 }),
+        (9, 9) => Some(Block::Water { level: 9 }),
+        (9, 10) => Some(Block::Water { level: 10 }),
+        (9, 11) => Some(Block::Water { level: 11 }),
+        (9, 12) => Some(Block::Water { level: 12 }),
+        (9, 13) => Some(Block::Water { level: 13 }),
+        (9, 14) => Some(Block::Water { level: 14 }),
+        (9, 15) => Some(Block::Water { level: 15 }),
```
Comment (terrarier2111) [inline blocks/src/versions/legacy.rs:71]: same as below

File: blocks/src/versions/legacy.rs [hunk 3]
```diff
@@ -0,0 +1,1692 @@
...
+        (55, 9) => Some(Block::RedstoneWire { east: RedstoneSide::None, south: RedstoneSide::None, north: RedstoneSide::None, west: RedstoneSide::None, power: 9 }),
+        (55, 10) => Some(Block::RedstoneWire { east: RedstoneSide::None, south: RedstoneSide::None, north: RedstoneSide::None, west: RedstoneSide::None, power: 10 }),
+        (55, 11) => Some(Block::RedstoneWire { east: RedstoneSide::None, south: RedstoneSide::None, north: RedstoneSide::None, west: RedstoneSide::None, power: 11 }),
+        (55, 12) => Some(Block::RedstoneWire { east: RedstoneSide::None, south: RedstoneSide::None, north: RedstoneSide::None, west: RedstoneSide::None, power: 12 }),
+        (55, 13) => Some(Block::RedstoneWire { east: RedstoneSide::None, south: RedstoneSide::None, north: RedstoneSide::None, west: RedstoneSide::None, power: 13 }),
+        (55, 14) => Some(Block::RedstoneWire { east: RedstoneSide::None, south: RedstoneSide::None, north: RedstoneSide::None, west: RedstoneSide::None, power: 14 }),
+        (55, 15) => Some(Block::RedstoneWire { east: RedstoneSide::None, south: RedstoneSide::None, north: RedstoneSide::None, west: RedstoneSide::None, power: 15 }),
```
Comment (terrarier2111) [inline blocks/src/versions/legacy.rs:373]: same as with lava

File: blocks/src/versions/legacy.rs [hunk 4]
```diff
@@ -0,0 +1,1692 @@
...
+        (51, 6) => Some(Block::Fire { east: false, south: false, north: false, west: false, up: false, age: 6 }),
+        (51, 7) => Some(Block::Fire { east: false, south: false, north: false, west: false, up: false, age: 7 }),
+        (51, 8) => Some(Block::Fire { east: false, south: false, north: false, west: false, up: false, age: 8 }),
+        (51, 9) => Some(Block::Fire { east: false, south: false, north: false, west: false, up: false, age: 9 }),
+        (51, 10) => Some(Block::Fire { east: false, south: false, north: false, west: false, up: false, age: 10 }),
+        (51, 11) => Some(Block::Fire { east: false, south: false, north: false, west: false, up: false, age: 11 }),
+        (51, 12) => Some(Block::Fire { east: false, south: false, north: false, west: false, up: false, age: 12 }),
+        (51, 13) => Some(Block::Fire { east: false, south: false, north: false, west: false, up: false, age: 13 }),
+        (51, 14) => Some(Block::Fire { east: false, south: false, north: false, west: false, up: false, age: 14 }),
+        (51, 15) => Some(Block::Fire { east: false, south: false, north: false, west: false, up: false, age: 15 }),
```
Comment (terrarier2111) [inline blocks/src/versions/legacy.rs:344]: same as with lava

File: blocks/src/versions/legacy.rs [hunk 5]
```diff
@@ -0,0 +1,1692 @@
...
+        (59, 6) => Some(Block::Wheat { age: 6 }),
+        (59, 7) => Some(Block::Wheat { age: 7 }),
+        (60, 0) => Some(Block::Farmland { moisture: 0 }),
+        (60, 1) => Some(Block::Farmland { moisture: 1 }),
+        (60, 2) => Some(Block::Farmland { moisture: 2 }),
+        (60, 3) => Some(Block::Farmland { moisture: 3 }),
+        (60, 4) => Some(Block::Farmland { moisture: 4 }),
+        (60, 5) => Some(Block::Farmland { moisture: 5 }),
+        (60, 6) => Some(Block::Farmland { moisture: 6 }),
+        (60, 7) => Some(Block::Farmland { moisture: 7 }),
```
Comment (terrarier2111) [inline blocks/src/versions/legacy.rs:392]: same as with lava

File: blocks/src/versions/legacy.rs [hunk 6]
```diff
@@ -0,0 +1,1692 @@
...
+        (77, 12) => Some(Block::StoneButton { powered: true, facing: Direction::North, face: AttachedFace::Wall }),
+        (77, 13) => Some(Block::StoneButton { powered: true, facing: Direction::South, face: AttachedFace::Floor }),
+        (78, 0) => Some(Block::Snow { layers: 1 }),
+        (78, 1) => Some(Block::Snow { layers: 2 }),
+        (78, 2) => Some(Block::Snow { layers: 3 }),
+        (78, 3) => Some(Block::Snow { layers: 4 }),
+        (78, 4) => Some(Block::Snow { layers: 5 }),
+        (78, 5) => Some(Block::Snow { layers: 6 }),
+        (78, 6) => Some(Block::Snow { layers: 7 }),
+        (78, 7) => Some(Block::Snow { layers: 8 }),
```
Comment (terrarier2111) [inline blocks/src/versions/legacy.rs:518]: same as with lava but with + 1

File: blocks/src/versions/legacy.rs [hunk 7]
```diff
@@ -0,0 +1,1692 @@
...
+        (81, 6) => Some(Block::Cactus { age: 6 }),
+        (81, 7) => Some(Block::Cactus { age: 7 }),
+        (81, 8) => Some(Block::Cactus { age: 8 }),
+        (81, 9) => Some(Block::Cactus { age: 9 }),
+        (81, 10) => Some(Block::Cactus { age: 10 }),
+        (81, 11) => Some(Block::Cactus { age: 11 }),
+        (81, 12) => Some(Block::Cactus { age: 12 }),
+        (81, 13) => Some(Block::Cactus { age: 13 }),
+        (81, 14) => Some(Block::Cactus { age: 14 }),
+        (81, 15) => Some(Block::Cactus { age: 15 }),
```
Comment (terrarier2111) [inline blocks/src/versions/legacy.rs:536]: same as with lava

File: src/model/liquid.rs [hunk 8]
```diff
@@ -34,18 +34,22 @@ pub fn render_liquid<W: Write>(
...
+        (63, 118, 228)
+    };
+
     let tex = match snapshot.get_block(x, y, z) {
         block::Block::Water { .. } => {
-            render::Renderer::get_texture(&textures, "minecraft:blocks/water_still")
-        }
-        block::Block::FlowingWater { .. } => {
-            render::Renderer::get_texture(&textures, "minecraft:blocks/water_flow")
+            render::Renderer::get_texture(&textures, "minecraft:block/water_still")
```
Comment (terrarier2111) [inline src/model/liquid.rs:46]: does this still work for older textures? or is that smth we should add in the future (support for older textures)

File: src/world/mod.rs [hunk 9]
```diff
@@ -137,37 +136,46 @@ impl World {
...
     pub fn update_block(&self, pos: Position) {
-        for yy in -1..2 {
-            for zz in -1..2 {
-                for xx in -1..2 {
-                    let bp = pos + (xx, yy, zz);
-                    let current = self.get_block(bp);
-                    let new = current.update_state(self, bp);
-                    if current != new {
-                        self.set_block_raw(bp, new);
+        if self.protocol_version < 404 {
```
Comment (terrarier2111) [inline src/world/mod.rs:139]: why don't we do this in later versions?

File: src/world/mod.rs [hunk 10]
```diff
@@ -137,37 +136,46 @@ impl World {
...
-                    let current = self.get_block(bp);
-                    let new = current.update_state(self, bp);
-                    let sky_light = self.get_sky_light(bp);
-                    let block_light = self.get_block_light(bp);
-                    if current != new {
-                        self.set_block_raw(bp, new);
-                        // Restore old lighting
-                        self.set_sky_light(bp, sky_light);
-                        self.set_block_light(bp, block_light);
+        if self.protocol_version < 404 {
```
Comment (terrarier2111) [inline src/world/mod.rs:159]: same as above

**Review by terrarier2111 — CHANGES_REQUESTED** [review]
Sorry for the delayed review and many repetitions at the very top.
note that i didn't look at the auto-genned files because that would have been an insane amount of work.

File: blocks/src/versions/legacy.rs (hunk 1 above)
Comment (nathanruiz) [reply blocks/src/versions/legacy.rs]: I've changed this to store `id & 0xF` in a variable before the match block, and then use it in these cases.

**Comment by nathanruiz** [comment]
@terrarier2111 I'm not sure how to fix the build issue. It seems like it's related to #255.

File: src/world/mod.rs (hunk 9 above)
Comment (terrarier2111) [reply src/world/mod.rs:139]: Could you add a comment explaining this rationale?

**Comment by terrarier2111** [comment]
> @terrarier2111 I'm not sure how to fix the build issue. It seems like it's related to #255.

I'll look into this in the upcoming days.

File: src/world/mod.rs (hunk 9 above)
Comment (nathanruiz) [reply src/world/mod.rs:143]: I've updated this

File: src/world/mod.rs (hunk 10 above)
Comment (nathanruiz) [reply src/world/mod.rs:169]: I've updated this

**Comment by terrarier2111** [comment]
So this should now be mergable, right?

**Pull request merged**